src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs

```diff
@@ -182,46 +183,22 @@ private async Task CreateConstantFieldActionsAsync(State state, ArrayBuilder<Cod
                 // If something is a constant, and it doesn't access any other locals constants,
                 // then we prefer to offer to generate a constant field instead of a constant
                 // local.
-                var action1 = CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action1, cancellationToken).ConfigureAwait(false))
+                if (CanGenerateIntoContainer(state, cancellationToken))
                 {
-                    actions.Add(action1);
+                    actions.Add(CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false));
                 }
 
-                var action2 = CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action2, cancellationToken).ConfigureAwait(false))
+                if (CanGenerateIntoContainer(state, cancellationToken))
                 {
-                    actions.Add(action2);
+                    actions.Add(CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false));
                 }
             }
         }
 
-        private async Task<bool> CanGenerateIntoContainerAsync(State state, CodeAction action, CancellationToken cancellationToken)
+        private bool CanGenerateIntoContainer(State state, CancellationToken cancellationToken)
         {
-            var result = await IntroduceFieldAsync(
-                state.Document, state.Expression,
-                allOccurrences: false, isConstant: state.IsConstant, cancellationToken: cancellationToken).ConfigureAwait(false);
-
-            var destination = result.Item2;
-            var insertionIndex = result.Item3;
-
-            if (!destination.OverlapsHiddenPosition(cancellationToken))
-            {
-                return true;
-            }
-
-            if (destination is TTypeDeclarationSyntax typeDecl)
-            {
-                var insertionIndices = GetInsertionIndices(typeDecl, cancellationToken);
-                if (insertionIndices != null &&
-                    insertionIndices.Count > insertionIndex &&
-                    insertionIndices[insertionIndex])
-                {
-                    return true;
-                }
-            }
```

Comment:
this is the core change.  Previously 'generate var' would say:

1. hey!  i'm trying to generate a constant.  let me see if i should generate a *field* constant instead of a *local* constant.
2. i can only generate a field if i can find a legal location for it.
3. to find a legal location, try to actually just generate the field then ensure the location we decided to put it in is 'legal'.
4. a position is 'legal' if it doesn't overlap a hidden region.

While this logic is fine, it's got high overhead because we both need to generate the field (expensive) and insist on this fine-grained approach of checking if the field dest we picked was not 'hidden'.

THe new logic is simply: 

1. generate a constant field as long as the containing type doesn't overlap a hidden region.

Technically this is more conservative as perhaps the user might have the case where the type overlaps a hidden region, but the exact insertion position happened to be non-hidden.  Note that this was totally a crapshoot *anyways* because it would depend entirely on which insertion position we picked (which totally depends on things like existing fields).  i.e. if there was an existing field at all in a hidden region (which is totally likely in scenarios with hidden code), then we wouldn't be able to offer the feature.

In general though, the ultra precision here (which again was a crapshoot) does not warrant the extreme cost of the work the code action needed to do.  Indeed, we should strive as much as possible to *not* have anlayzers/refactorings actually go an *execute* their expensive code-actions to determine if they should offer their code-actions.

Reply:
Note: i looked into seeing if we could just extract hte 'determine insertion location' logic independent of everything else (to keep all scenarios working) and it was just so complex (and still expensive).  That's because the 'determine insertion location' logic ends up doing full walks and rewrites of the tree to see what parts of the code are edited to drive the insertion logic based on that.

i.e. it will say "hey... even if i wanted to insert the field down bwlow, i'll choose a different location based on what was actually updated in the user case to fixup references to that constant".

Reply:
IMO, this complexity is *not* worth it, and it's fine to say "if the containing type overlaps hidden code, we don't offer to generate constant fields".  Alternatively, we could flip things around and say: as long as there's a valid insertion position in the containing type, we will allow generation.  And we will drive the actual destination location based on that.  I'd be fine with adding that complexity, though i would argue it's just not worth it.

Reply:
Alternative approach is here @sharwell https://github.com/dotnet/roslyn/pull/40077